### PR TITLE
Saas 12 smallfixes3 jco sq

### PR DESCRIPTION
**Pull request by jco-odoo**
See details in each commit
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/maintenance/models/maintenance.py [hunk 1]
```diff
@@ -143,7 +145,7 @@ def name_search(self, name, args=None, operator='ilike', limit=100):
     maintenance_team_id = fields.Many2one('maintenance.team', string='Maintenance Team')
     maintenance_duration = fields.Float(help="Maintenance Duration in minutes and seconds.")
 
-    @api.depends('period', 'maintenance_open_count', 'maintenance_ids.request_date')
+    @api.depends('period', 'maintenance_ids', 'maintenance_ids.request_date')
```
Comment (tde-banana-odoo) [inline addons/maintenance/models/maintenance.py]: Actually correct trigger is maintenance_ids.request_date

File: addons/mrp/models/mrp_workorder.py [hunk 2]
```diff
@@ -320,7 +320,11 @@ def record_production(self):
                                      'workorder_id': self.id,
                                      })
             else:
-                production_move.quantity_done += self.qty_producing  # TODO: UoM conversion?
+                production_move.quantity_done += self.qty_producing
```
Comment (tde-banana-odoo) [inline addons/mrp/models/mrp_workorder.py]: Isn't it a module to handle by products ?

File: addons/maintenance/models/maintenance.py [hunk 3]
```diff
@@ -202,17 +204,53 @@ def _read_group_category_ids(self, domain, read_group_order=None, access_rights_
             fold[category.id] = category.fold
         return result, fold
 
+    def _create_new_request(self, date):
+        self.ensure_one()
+        self.env['maintenance.request'].create({
+                    'name': _('Preventive Maintenance - %s') % self.name,
+                    'request_date': date.strftime(DEFAULT_SERVER_DATE_FORMAT),
+                    'schedule_date': date.strftime(DEFAULT_SERVER_DATE_FORMAT),
```
Comment (xmo-odoo) [inline addons/maintenance/models/maintenance.py]: `fields.Date.to_string`?

File: addons/maintenance/models/maintenance.py [hunk 4]
```diff
@@ -202,17 +204,53 @@ def _read_group_category_ids(self, domain, read_group_order=None, access_rights_
...
+                    'category_id': self.category_id.id,
+                    'equipment_id': self.id,
+                    'maintenance_type': 'preventive',
+                    })
+
+    def _fill_up(self, from_date, to_date, to_end=False):
+        """
+            to_end takes into account that the distance to to_end is period also
```
Comment (xmo-odoo) [inline addons/maintenance/models/maintenance.py]: what?

File: addons/maintenance/models/maintenance.py [hunk 5]
```diff
@@ -202,17 +204,53 @@ def _read_group_category_ids(self, domain, read_group_order=None, access_rights_
...
+        """
+        self.ensure_one()
+        period = self.period
+        current_date = from_date + relativedelta(days=period)
+        while (current_date < to_date):
+            if to_end or current_date + relativedelta(days=period) < to_date:
+                self._create_new_request(current_date)
+            current_date = current_date + relativedelta(days=period)
```
Comment (xmo-odoo) [inline addons/maintenance/models/maintenance.py]: Couldn't that use `dateutil.rrule.rrule`? It works pretty well to iterate on dates.

File: addons/maintenance/models/maintenance.py [hunk 6]
```diff
@@ -202,17 +204,53 @@ def _read_group_category_ids(self, domain, read_group_order=None, access_rights_
...
-                })
+        show_until = (fields.Datetime.from_string(fields.Datetime.now()) + relativedelta(months=1)).strftime(DEFAULT_SERVER_DATE_FORMAT)
+        date_now = fields.Datetime.now()
+        for equipment in self.search([('period', '>', 0), ('next_action_date', '<', show_until)]):
+            requests = self.env['maintenance.request'].search([('request_date', '<', show_until), 
+                                                               ('stage_id.done', '=', False), 
+                                                               ('equipment_id', '=', equipment.id),
+                                                               ('maintenance_type', '=', 'preventive')],
+                                                              order='request_date') #('request_date', '>', date_now),
+            request_dates = [x.request_date for x in requests]
```
Comment (xmo-odoo) [inline addons/maintenance/models/maintenance.py]: `requests.mapped('request_date')`.

File: addons/mrp/models/mrp_workorder.py [hunk 7]
```diff
@@ -320,7 +320,11 @@ def record_production(self):
...
+            #by_products in the end (could be improved by using operation_id on by_products)
+            for move in self.production_id.move_finished_ids.filtered(lambda x: x.id != production_move.id and x.state not in ('done', 'cancel')):
+                if move.product_id.tracking == 'none':
```
Comment (xmo-odoo) [inline addons/mrp/models/mrp_workorder.py]: why not also put that in the filtered?

File: addons/maintenance/models/maintenance.py [hunk 8]
```diff
@@ -202,17 +204,53 @@ def _read_group_category_ids(self, domain, read_group_order=None, access_rights_
...
+                if date >= date_now:
+                    if last_date:
+                        equipment._fill_up(fields.Datetime.from_string(last_date), fields.Datetime.from_string(date))
+                    elif fields.Datetime.from_string(date) >= fields.Datetime.from_string(equipment.next_action_date) + relativedelta(days=equipment.period):
+                        equipment._fill_up(first_date, fields.Datetime.from_string(date))
+                last_date=date
+            if not last_date:
+                last_date = fields.Datetime.now()
+            if last_date < show_until:
+                equipment._fill_up(fields.Datetime.from_string(last_date), fields.Datetime.from_string(show_until), to_end=True)
```
Comment (xmo-odoo) [inline addons/maintenance/models/maintenance.py]: `last_date` can be a datetime here. Maybe just make all the request_dates into datetimes out of the box or something?

**Comment by tde-banana-odoo** [comment]
Work continues on https://github.com/odoo/odoo/pull/14368 .

**Pull request closed without merging**